File: Div3-Group1/MergeTwoSortedLists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {
        ListNode *temp=new ListNode();
        ListNode *tm=temp;

        while (list1 != NULL && list2 != NULL)
        {
            if (list1->val < list2->val)
            {
                ListNode * x=new ListNode(list1->val);
                tm->next=x;
                tm=tm->next;
                list1=list1->next;
            }
            else
            {
                ListNode * x=new ListNode(list2->val);
                tm->next=x;
                tm=tm->next;
                list2=list2->next;

            }
        }
        while (list1!=NULL)
        {
            ListNode * x=new ListNode(list1->val);
            tm->next=x;
            tm=tm->next;
            list1=list1->next;
        }
        while (list2!=NULL)
        {
            ListNode * x=new ListNode(list2->val);
            tm->next=x;
            tm=tm->next;
            list2=list2->next;
        }
        
        return temp->next;

    }
};

File: leet-code-solutions/leetcode/merge-two-sorted-lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        if not list1:
            return list2
        if not list2:
            return list1
        
        if not list2 and not list1:
            return list1
        
        head=ListNode()
        temp=head

        temp1=list1
        temp2=list2

        while temp1 and temp2:
            if temp1 and temp2: 
                if temp1.val <= temp2.val:
                    temp.next=temp1
                    temp = temp.next
                    temp1 = temp1.next
                else:
                    temp.next=temp2
                    temp = temp.next
                    temp2 = temp2.next
        
        while temp1:
            temp.next=temp1
            temp = temp.next
            temp1 = temp1.next

        while temp2:
            temp.next=temp2
            temp = temp.next
            temp2 = temp2.next
            
        return head.next
                
